README.md
# filekv
This is a key&value storage library which uses file system to store data.

History.md
1.0.9 / 2015-07-11
==================
  * Added events listener : add event; set event; add event; replace event; 
  * Changed read line begin #1 line, but not #0 line
  * Use All-New readfileline 2.0.2
  * Added appveyor ci

1.0.0 / 2015-07-02
==================

  * Init


test/test.js
"use strict";


var now = Date.now();
var util = require('util');
var fs = require('fs');
var assert = require('assert')
var fileDir = __dirname+'/filekv_test_data_folder_'+now;
var fkvObj = require('../index.js').create({
	fileDir:fileDir,
	workMax:1000
});

assert.equal(typeof fkvObj,'object');
assert.notEqual(fkvObj,null);
assert.notEqual(fkvObj,undefined);

var value = 'filekv test string';


fkvObj.on('set',function(input,output){

	var key = input[0];
	console.log('set event emit: '+key);
});

fkvObj.on('get',function(input,output){
	var key = input[0];
	console.log('get event emit: '+key);
	var err = output[0];
	var outValue = output[1];
	switch(key){
		case 'testkey1':
		case 'testkey2':
		case 'testkey3':
			assert.equal(outValue==value || !outValue,true);
			break;
	}
});

fkvObj.on('replace',function(input,output){
	var key = input[0];
	var val = input[1];
	console.log('replace event emit: '+key);

	if(input[0]=='testkey5-2'){
		//have error
		assert.equal(output[0] instanceof Error,true);
		fkvObj.get(key,function(err,data){
			assert.notEqual(val,data);
		});
	}
	if(input[0]=='testkey5'){
		//no error
		assert.equal(output[0] instanceof Error,false);
		fkvObj.get(key,function(err,data){
			assert.equal(val,data);
		});
	}
});
fkvObj.on('add',function(input,output){
	var key = input[0];
	var val = input[1];
	console.log('add event emit: '+key);

	if(input[0]=='testkey4-2'){
		//no error
		assert.equal(output[0] instanceof Error,false);
		fkvObj.get(key,function(err,data){
			assert.equal(val,data);
		});

	}
	if(input[0]=='testkey4'){

		//have error
		assert.equal(output[0] instanceof Error,true);
		fkvObj.get(key,function(err,data){
			assert.notEqual(val,data);
		});

	}
});

fkvObj.on('delete',function(input,output){
	var key = input[0];
	console.log('delete event emit: '+key);
	fkvObj.get(key,function(err,data){

		assert.equal(err instanceof Error,true);
		assert.notEqual(data,value);

	});
});

//test set, and, get
fkvObj.set('testkey1',value,function(err){
	fkvObj.get('testkey1',function(err,data,info){

		assert.equal(typeof data,'string');
		assert.equal(data,value);

	})
})

//test set, delete, get
fkvObj.set('testkey2',value,function(err){
	fkvObj.del('testkey2',function(){
		fkvObj.get('testkey2',function(err,data){

			assert.equal(!!data,false);
			assert.notEqual(data,value);

		})		
	});
})

//test key life time
fkvObj.set('testkey3',value,5,function(){
	// assert.equal(1,2)
	fkvObj.get('testkey3',function(err,data,info){

		assert.equal(value,data);
		assert.equal(typeof value,typeof data);

	})
	setTimeout(function(){
		fkvObj.get('testkey3',function(err,data){
			assert.equal(value,data);
			assert.equal(typeof value,typeof data);

		})
	},3e3);

	setTimeout(function(){
		fkvObj.get('testkey3',function(err,data){
			assert.notEqual(value,data);
			assert.notEqual(typeof value,typeof data);
			assert.equal(!!data,false);

		})
	},5e3);

	setTimeout(function(){
		fkvObj.get('testkey3',function(err,data){
			assert.notEqual(value,data);
			assert.notEqual(typeof value,typeof data);
			assert.equal(!!data,false);
		})
	},7e3);
})


//test add function
fkvObj.set('testkey4',value,function(){
	fkvObj.add('testkey4',value+'add',function(err){
		assert.equal(err instanceof Error,true);
	})
	fkvObj.add('testkey4-2',value+'add4-2',function(err){
		assert.equal(err,null);
	})

});

//test replace function
fkvObj.set('testkey5',value,function(){
	fkvObj.replace('testkey5',value+'replace',function(err){
		
		assert.equal(err,null);		
	})
	fkvObj.replace('testkey5-2',value+'replace5-2',function(err){
		assert.equal(err instanceof Error,true);
	})

});



/*
 * lib md5 testing
 */
var _md5 = require('../lib/crypto.js').md5;
assert.equal(_md5('123').toUpperCase(),'202cb962ac59075b964b07152d234b70'.toUpperCase());
assert.equal(_md5('filekv').toUpperCase(),'ec9310957ee4cedd23ce617051c58ea3'.toUpperCase());
assert.equal(_md5('中文').toUpperCase(),'a7bac2239fcdcb3a067903d8077c4a07'.toUpperCase());


/*
 * lib tool testing
 */
var tool = require('../lib/tool.js');
var nowMd5 = _md5(now);
assert.equal(tool.buildDataFileSubDir(nowMd5),nowMd5.substr(0,3)+'/'+nowMd5.substr(3,3));


/*
 * lib mkdirs testing
 */

var testDataFolder = fileDir+'/'+tool.buildDataFileSubDir(_md5(Date.now()));

var filekvFSTool = require('../lib/fs.js');
fs.unlink(testDataFolder,function(err){
	
	filekvFSTool.mkdirs(testDataFolder,function(err){
		
		fs.exists(testDataFolder,function(isExsts){
			assert.equal(isExsts,true);
		});
	});
});



test/test_2.js
"use strict";


var now = Date.now();
var util = require('util');
var fs = require('fs');
var assert = require('assert')
var fileDir = __dirname+'/filekv_test_data_folder_'+now;
var fkvObj = require('../index.js').create({
	fileDir:fileDir,
	workMax:1000
});


fkvObj.set('test','test value',function(err,data){
	console.log(err,data);
	fkvObj.get('test',function(err,data,info){
		console.log(err,data,info);
	})

});

lib/filekv.js
/**
 * User:	zenboss
 * GitHub:	zenboss
 * Date:	2015-07-02
 * Time:	09:54 A.M.
 * Email:	zenyes@gmail.com
 */
"use strict";
var readFileLine	= require('readfileline');
var workQueue		= require('workqueue');

var fs = require('fs');

var events = require('events');
var util = require('util');

var _md5 = require('./crypto.js').md5;
var fkvtool = require('./tool.js');

var _innerNextTick = setImmediate;



var filekv = function(config){
	events.EventEmitter.call(this);

	var self = this;
	self.setFileDir(config.fileDir);

	self._workQueue = workQueue.create({
		workMax:config.workMax||config.workQueueMax||1000
	});

};
util.inherits(filekv,events.EventEmitter);

filekv.create = function(config){
	return new filekv(config);
};

filekv.prototype.fs = require('./fs.js');


filekv.prototype.setFileDir = function(path){
	var self = this;
	if(!path){
		throw new Error("Config item:\"fileDir\" can't empty, which not have default value!");
		return;
	}
	self.fileDir = path;
	return self.fileDir;
};
filekv.prototype.setWorkMax = function(maxnum){
	this._workQueue.setQueueMax(maxnum);
}



filekv.prototype._get = function(key,cb,opt){
	var self = this;

	cb = cb||function(){};
	opt = opt||{};

	var md5key = _md5(key);
	var filePath = this.fileDir+'/'+fkvtool.buildDataFileSubDir(md5key)+'/'+md5key+'.fkv';
	var valueData = null;
	var updateTime = 0;
	var expireTime = 0;

	self._workQueue.queue(function(queueCB){

		//this function is unify callback function, it do many job.
		var unifyCallBack = function(err,valueData,info){ 
			queueCB.apply(self,arguments);
			cb.apply(self,arguments);
			
		}

		readFileLine(filePath,function(lineData,lineNum){
			var theRead = this;

			switch(lineNum){
				case 1:

					expireTime = parseInt(lineData+'')||0;
					if(expireTime!=0 && expireTime<=parseInt(Date.now()/1000)){
						self._del(key);
						var cbError = new Error('key expired');
						unifyCallBack(cbError);
						theRead.close();
					}
					break;
				case 2:
					updateTime = parseInt(lineData+'')||0;
					break;
				case 3: //data
					try{
						valueData = JSON.parse(lineData+'', function(key, value) {
		    					return value && value.type === 'Buffer' 
		    					? new Buffer(value.data)
		      					: value; // Buffer类型特殊处理
		  				});
					}catch(ex){
						unifyCallBack(ex);
						theRead.close();
					}
					break;
				
			}
		},function(err,endType,totalLineNum){
			if(err){
				unifyCallBack(err);
			}else{
				if(endType=='end'){

					if(updateTime>0 && totalLineNum>=3){
						unifyCallBack(null,valueData,{
							updateTime:updateTime,
							expireTime:expireTime,
							filePath:filePath
						});
					}else{
						unifyCallBack(new Error("file content can't parse"));
					}

				}
			}
		});

	});
	
};


filekv.prototype._set = function(key,value,lifeTime,cb,opt){
	var self = this;
	if('function' == typeof lifeTime){
		cb = lifeTime;
		lifeTime = 0;
	}

	opt 		= opt||{};
	cb 			= cb||function(){};
	lifeTime	= parseInt(lifeTime)||0;

	var md5key		= _md5(key);
	var filePath	= this.fileDir+'/'+fkvtool.buildDataFileSubDir(md5key)+'/';


	self._workQueue.queue(function(queueCB){
		var unifyCallBack = function(err,_more){
			queueCB(err,_more);
			cb(err,_more);
		}
		self.fs.mkdirs(filePath,function(err,_more){
			if(!!err){
				unifyCallBack(err,_more);
				return;
			}
			var fileAllPath		= filePath+'/'+md5key+'.fkv';
			var updateTime		= parseInt(Date.now()/1000);
			var expireTime		= 0;

			if(lifeTime!=0)expireTime = updateTime + lifeTime;

			var fileData = '';
			fileData += expireTime+'\n';
			fileData += updateTime+'\n';
			fileData += JSON.stringify(value);
				
			fs.writeFile(fileAllPath,fileData,function(err,_more){
				unifyCallBack(err,_more);
				if(expireTime!=0 && expireTime<=updateTime){
					self._del(key);
				}
			});

		});

	});
};


filekv.prototype._del = function(key,cb,opt){
	var self = this;
	if('function' == typeof opt){
		cb = opt;
		opt = {};
	}
	cb = cb||function(){};

	var md5key = _md5(key);
	var filePath = this.fileDir+'/'+fkvtool.buildDataFileSubDir(md5key)+'/'+md5key+'.fkv';
	fs.unlink(filePath,cb);

};




filekv.prototype.has = function(key,cb,opt){
	var self = this;

	cb = cb||function(){};
	opt = opt || {};
	var md5key = _md5(key);
	var filePath = this.fileDir+'/'+fkvtool.buildDataFileSubDir(md5key)+'/'+md5key+'.fkv';
	var expireTime = 0;
	self._workQueue.queue(function(queueCB){

		var unifyCallBack = function(){
			if(expireTime!=0 && expireTime<=parseInt(Date.now()/1000)){
				cb(new Error('key expired'),false);
			}else{
				cb(null,filePath);
			}
			queueCB(null,filePath);
		}

		readFileLine(filePath,function(lineData,lineNum){
			var theRead = this;
			switch(lineNum){
				case 1:
					expireTime = parseInt(lineData)||0;
					unifyCallBack();
					theRead.close();
					return false;
					break;
				default:
					theRead.close();
					return false;
			}

		},function(err,endType,nowLineNum){
			if(err){
				cb(err,false);
			}else{
				if(endType=='end'){
					unifyCallBack();
				}
			}
		});
	});
};

filekv.prototype.add = function(key,value,lifeTime,cb,opt){
	var self = this;
	var callArgs = arguments;
	
	if('function' == typeof lifeTime){
		cb = lifeTime;
		lifeTime = 0;
	}
	cb = cb || function(){};

	var unifyCB = function(){
		cb.apply(self,arguments);
		self.emit('add',callArgs,arguments);
	};
	self.has(key,function(err,isExist){
		if(isExist){
			unifyCB(new Error('key existed'));
		}else{
			self._set(key,value,lifeTime,unifyCB,opt);
		}
	})
};

filekv.prototype.replace = function(key,value,lifeTime,cb,opt){
	var self = this;
	var callArgs = arguments;

	if('function' == typeof lifeTime){
		cb = lifeTime;
		lifeTime = 0;
	}
	cb = cb || function(){};

	var unifyCB = function(){
		cb.apply(self,arguments);
		self.emit('replace',callArgs,arguments);
	};
	self.has(key,function(err,isExist){
		if(!isExist){
			unifyCB(new Error('key not exist'));
		}else{
			self._set(key,value,lifeTime,unifyCB,opt);
		}
	})
};

filekv.prototype.get = function(key,cb,opt){
	var self = this;
	var callArgs = arguments;
	cb = cb || function(){};
	self._get.call(self,key,function(){
		cb.apply(self,arguments);
		self.emit('get',callArgs,arguments);
	},opt);
};

filekv.prototype.set = function(key,value,lifeTime,cb,opt){
	var self = this;
	var callArgs = arguments;
	if('function' == typeof lifeTime){
		cb = lifeTime;
		lifeTime = 0;
	}
	cb = cb || function(){};
	lifeTime = lifeTime||0;
	self._set.call(self,key,value,lifeTime,function(){
		cb.apply(self,arguments);
		self.emit('set',callArgs,arguments);
	},opt);
};


filekv.prototype.delete = filekv.prototype.del =function(key,cb,opt){
	var self = this;
	var callArgs = arguments;
	cb = cb || function(){};
	self._del.call(self,key,function(){
		cb.apply(self,arguments);
		self.emit('delete',callArgs,arguments);
	},opt);
}



module.exports = filekv;
